File: app/src/main/java/ru/samtakoy/core/screens/courses/CoursesAdapter.java
package ru.samtakoy.core.screens.courses;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import ru.samtakoy.R;
import ru.samtakoy.core.model.LearnCourse;

public class CoursesAdapter extends RecyclerView.Adapter <CoursesAdapter.CoursesItemHolder>{

    public interface CourseClickListener{
        void onCourseClick(LearnCourse course);
    }

    class CoursesItemHolder extends RecyclerView.ViewHolder implements View.OnClickListener{

        private LearnCourse mCourse;
        private TextView mText;
        private CourseClickListener mCourseClickListener;

        public CoursesItemHolder(@NonNull View itemView, CourseClickListener clickListener) {
            super(itemView);
            mCourseClickListener = clickListener;
            mText = itemView.findViewById(R.id.list_item_text);
            itemView.setOnClickListener(this);
        }

        public void bindCourse(LearnCourse t){
            mCourse = t;
            mText.setText(t.getDynamicTitle());

        }

        @Override
        public void onClick(View view) {
            mCourseClickListener.onCourseClick(mCourse);
        }

    }

    private CourseClickListener mCourseClickListener;
    private List<LearnCourse> mCurCourses;

    public CoursesAdapter(CourseClickListener clickListener){
        mCurCourses = new ArrayList<>();
        mCourseClickListener = clickListener;
    }

    public void setCurCourses(List<LearnCourse> curCourses) {
        mCurCourses = curCourses;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public CoursesItemHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater li = LayoutInflater.from(parent.getContext());
        return new CoursesItemHolder(
                li.inflate(R.layout.courses_list_item, parent, false),
                mCourseClickListener
        );
    }

    @Override
    public void onBindViewHolder(@NonNull CoursesItemHolder holder, final int position) {
        holder.bindCourse(mCurCourses.get(position));
        holder.itemView.setOnCreateContextMenuListener(null);
        holder.itemView.setOnLongClickListener(null);
    }

    @Override
    public int getItemCount() {
        return mCurCourses.size();
    }

    @Override
    public void onViewRecycled(@NonNull CoursesItemHolder holder) {
        holder.itemView.setOnLongClickListener(null);
        super.onViewRecycled(holder);
    }

}

File: app/src/main/java/ru/samtakoy/core/presentation/courses/model/CoursesAdapter.kt
package ru.samtakoy.core.presentation.courses.model

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import ru.samtakoy.R
import ru.samtakoy.core.presentation.courses.model.CoursesAdapter.CoursesItemHolder

class CoursesAdapter(
    private val clickListener: CourseClickListener
) : RecyclerView.Adapter<CoursesItemHolder>() {

    fun interface CourseClickListener {
        fun onCourseClick(item: CourseItemUiModel)
    }

    inner class CoursesItemHolder(
        itemView: View,
        private val mCourseClickListener: CourseClickListener
    ) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        private var mCourse: CourseItemUiModel? = null
        private val mText: TextView

        init {
            mText = itemView.findViewById<TextView>(R.id.list_item_text)
            itemView.setOnClickListener(this)
        }

        fun bindCourse(data: CourseItemUiModel) {
            mCourse = data
            mText.setText(data.title)
        }

        override fun onClick(view: View?) {
            mCourse?.let {
                mCourseClickListener.onCourseClick(it)
            }
        }
    }

    private var mCurCourses: List<out CourseItemUiModel>

    init {
        mCurCourses = ArrayList<CourseItemUiModel>()
    }

    fun setCurCourses(curCourses: List<out CourseItemUiModel>) {
        mCurCourses = curCourses
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): CoursesItemHolder {
        val li = LayoutInflater.from(parent.getContext())
        return CoursesItemHolder(
            li.inflate(R.layout.courses_list_item, parent, false),
            clickListener
        )
    }

    override fun onBindViewHolder(holder: CoursesItemHolder, position: Int) {
        holder.bindCourse(mCurCourses.get(position))
        holder.itemView.setOnLongClickListener(null)
    }

    override fun getItemCount(): Int {
        return mCurCourses.size
    }

    override fun onViewRecycled(holder: CoursesItemHolder) {
        holder.itemView.setOnLongClickListener(null)
        super.onViewRecycled(holder)
    }
}
